script/v2/manual-runs/scWETHv2Rebalance.s.sol

```diff
@@ -0,0 +1,392 @@
+    //         adaptersToInvest.push(morphoAdapter);
+    //         allocationPercents.push(MORPHO_ALLOCATION_PERCENT);
+    //         targetLtv[morphoAdapter] = MORPHO_TARGET_LTV;
+
+    //         totalAllocationPercent += MORPHO_ALLOCATION_PERCENT;
+    //     }
+
+    //     if (AAVEV3_ALLOCATION_PERCENT > 0) {
+    //         adaptersToInvest.push(aaveV3Adapter);
+    //         allocationPercents.push(AAVEV3_ALLOCATION_PERCENT);
+    //         targetLtv[aaveV3Adapter] = AAVEV3_TARGET_LTV;
+
+    //         totalAllocationPercent += AAVEV3_ALLOCATION_PERCENT;
+    //     }
+
+    //     if (COMPOUNDV3_ALLOCATION_PERCENT > 0) {
+    //         adaptersToInvest.push(compoundV3Adapter);
+    //         allocationPercents.push(COMPOUNDV3_ALLOCATION_PERCENT);
+    //         targetLtv[compoundV3Adapter] = COMPOUNDV3_TARGET_LTV;
+
+    //         totalAllocationPercent += COMPOUNDV3_ALLOCATION_PERCENT;
+    //     }
+
+    //     require(totalAllocationPercent == 1e18, "totalAllocationPercent != 100%");
+    // }
+
+    struct RebalanceDataParams {
+        IAdapter adapter;
+        uint256 amount;
+        uint256 flashLoanAmount;
+        bool isSupplyBorrow;
+    }
+
+    RebalanceDataParams[] rebalanceDataParams;
+
+    function _investAmount() internal view returns (uint256 investAmount) {
+        uint256 float = weth.balanceOf(address(vault));
+        uint256 minimumFloat = vault.minimumFloatAmount();
+        // investAmount == 0 just reinvests profits
+        investAmount = float > minimumFloat ? float - minimumFloat : 0;
+    }
+
+    function _updateRebalanceParams() internal {
+        uint256 investAmount = _investAmount();
+
+        IAdapter[3] memory allAdapters = [morphoAdapter, compoundV3Adapter, aaveV3Adapter];
+
+        IAdapter adapter;
+        uint256 allocationPercent;
+        uint256 amount;
+        uint256 flashLoanAmount;
+        bool isSupplyBorrow;
+        for (uint256 i; i < allAdapters.length; i++) {
+            adapter = allAdapters[i];
+            allocationPercent = adapterAllocationPercent[adapter];
+
+            if (allocationPercent != 0) {
+                amount = investAmount.mulWadDown(allocationPercent);
+                // first check if allocation Percent is greater than zero
+                (flashLoanAmount, isSupplyBorrow) = _calcRebalanceFlashLoanAmount(adapter, amount);
+
+                rebalanceDataParams.push(RebalanceDataParams(adapter, amount, flashLoanAmount, isSupplyBorrow));
+            } else {
+                uint256 ltv = getLtv(adapter);
+                // even if there is no allocation here, we still want to check if a disinvest is needed in this protocol
+                if (ltv > targetLtv[adapter] + disinvestThreshold) {
+                    (flashLoanAmount, isSupplyBorrow) = _calcRebalanceFlashLoanAmount(adapter, 0);
+
+                    rebalanceDataParams.push(RebalanceDataParams(adapter, 0, flashLoanAmount, false));
+                }
+            }
```

Comment:
should there be a check if an adapter is actually spoorted?

Reply:
I dont think that is needed. Since the contract will revert if an unsupported adapter is added here anyway. And the adapter addresses are fixed in the constants file.

Reply:
was thinking that there is no point wasting gas by letting this fail at execution time because of that

Reply:
[Done](https://github.com/lindy-labs/sandclock-contracts/pull/127/commits/b1b3b1998e68ba4d5c5fad93202625fd2da5b6bb)

Reply:
I think we need to add another test for the case when not all of the adapters are available on the vault.
Setup the test so the vault supports only 2 adapters (remove one to achieve this) and make sure the script is not going to fail.

Reply:
[Added](https://github.com/lindy-labs/sandclock-contracts/pull/127/commits/f5233a403c680f2ecf4721201be0ab1d57540abc)